File: ditto-go-client/http_client.go
package ditto

import (
	"bytes"
	"crypto/tls"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"
)

type HTTPClientOptions struct {
	Host               string
	Port               int
	TLS                bool
	Username           string
	Password           string
	RejectUnauthorized bool
	InsecureSkipVerify bool
	DevInsecureTLS     bool
	TrustedCertPath    string
	ConnectTimeout     time.Duration
	RequestTimeout     time.Duration
	Timeout            time.Duration
	StrictMode         bool
}

type HTTPClient struct {
	baseURL    string
	httpClient *http.Client
	authHeader string
	strictMode bool
}

const insecureTLSDevEnvGate = "DITTO_CLIENT_ALLOW_INSECURE_TLS_DEV"

func NewHTTPClient(opts HTTPClientOptions) *HTTPClient {
	host := opts.Host
	if host == "" {
		host = "localhost"
	}
	port := opts.Port
	if port == 0 {
		port = 7778
	}
	scheme := "http"
	if opts.TLS {
		scheme = "https"
	}
	connectTimeout := opts.ConnectTimeout
	if connectTimeout <= 0 {
		connectTimeout = opts.Timeout
	}
	if connectTimeout <= 0 {
		connectTimeout = 10 * time.Second
	}
	requestTimeout := opts.RequestTimeout
	if requestTimeout <= 0 {
		requestTimeout = opts.Timeout
	}
	if requestTimeout <= 0 {
		requestTimeout = 10 * time.Second
	}
	tr := &http.Transport{
		DialContext: (&net.Dialer{
			Timeout: connectTimeout,
		}).DialContext,
	}
	if opts.TLS {
		if opts.InsecureSkipVerify || opts.DevInsecureTLS {
			if allowInsecureTLSDev() {
				log.Printf(
					"WARNING: insecure TLS verification is enabled for development because %s is set; do not use this in production",
					insecureTLSDevEnvGate,
				)
			} else {
				log.Printf(
					"WARNING: insecure TLS bypass flags require %s=true and will otherwise be ignored; use a trusted certificate configuration instead",
					insecureTLSDevEnvGate,
				)
			}
		}
		tlsConfig := &tls.Config{InsecureSkipVerify: allowInsecureTLSDev() && (opts.InsecureSkipVerify || opts.DevInsecureTLS)}
		if strings.TrimSpace(opts.TrustedCertPath) != "" {
			pemBytes, err := os.ReadFile(strings.TrimSpace(opts.TrustedCertPath))
			if err != nil {
				panic(fmt.Sprintf("failed to read trusted cert path %q: %v", opts.TrustedCertPath, err))
			}
			pool, err := x509.SystemCertPool()
			if err != nil || pool == nil {
				pool = x509.NewCertPool()
			}
			if !pool.AppendCertsFromPEM(pemBytes) {
				panic(fmt.Sprintf("failed to parse trusted cert path %q", opts.TrustedCertPath))
			}
			tlsConfig.RootCAs = pool
		}
		tr.TLSClientConfig = tlsConfig
	}
	c := &HTTPClient{
		baseURL: fmt.Sprintf("%s://%s:%d", scheme, host, port),
		httpClient: &http.Client{
			Timeout:   requestTimeout,
			Transport: tr,
		},
		strictMode: opts.StrictMode,
	}
	if opts.Username != "" && opts.Password != "" {
		c.authHeader = "Basic " + base64.StdEncoding.EncodeToString([]byte(opts.Username+":"+opts.Password))
	}
	return c
}

func (c *HTTPClient) Close() {}

func (c *HTTPClient) request(method, path string, body []byte, contentType string, headers map[string]string) ([]byte, int, error) {
	req, err := http.NewRequest(method, c.baseURL+path, bytes.NewReader(body))
	if err != nil {
		return nil, 0, err
	}
	if c.authHeader != "" {
		req.Header.Set("Authorization", c.authHeader)
	}
	if contentType != "" {
		req.Header.Set("Content-Type", contentType)
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, 0, err
	}
	defer resp.Body.Close()
	b, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, resp.StatusCode, err
	}
	return b, resp.StatusCode, nil
}

func httpStatusToCode(status int) string {
	switch status {
	case 503:
		return ErrNodeInactive
	case 504:
		return ErrWriteTimeout
	case 404:
		return ErrKeyNotFound
	default:
		return ErrInternalError
	}
}

func parseHTTPError(status int, body []byte) error {
	if status >= 200 && status < 300 {
		return nil
	}
	msg := string(body)
	code := httpStatusToCode(status)
	var payload struct {
		Error   string `json:"error"`
		Message string `json:"message"`
	}
	if json.Unmarshal(body, &payload) == nil {
		if payload.Error != "" {
			code = payload.Error
		}
		if payload.Message != "" {
			msg = payload.Message
		} else if payload.Error != "" {
			msg = payload.Error
		}
	}
	return &DittoError{Code: code, Message: msg}
}

func (c *HTTPClient) Ping() (bool, error) {
	b, status, err := c.request(http.MethodGet, "/ping", nil, "", nil)
	if err != nil {
		return false, err
	}
	if status != http.StatusOK {
		return false, nil
	}
	var p struct {
		Pong bool `json:"pong"`
	}
	if err := json.Unmarshal(b, &p); err != nil {
		return false, err
	}
	return p.Pong, nil
}

func (c *HTTPClient) Get(key string, namespace ...string) (*GetResult, error) {
	ns, err := resolveCoreNamespace(c.strictMode, "get", key, namespace...)
	if err != nil {
		return nil, err
	}
	b, status, err := c.request(http.MethodGet, "/key/"+url.PathEscape(key), nil, "", namespaceHeaderPtr(ns))
	if err != nil {
		return nil, err
	}
	if status == http.StatusNotFound {
		return nil, nil
	}
	if err := parseHTTPError(status, b); err != nil {
		return nil, err
	}
	var payload struct {
		Value       string `json:"value"`
		ValueBase64 string `json:"value_base64"`
		Version     uint64 `json:"version"`
	}
	if err := json.Unmarshal(b, &payload); err != nil {
		return nil, err
	}
	value := []byte(payload.Value)
	if payload.ValueBase64 != "" {
		decoded, err := base64.StdEncoding.DecodeString(payload.ValueBase64)
		if err != nil {
			return nil, err
		}
		value = decoded
	}
	return &GetResult{Value: value, Version: payload.Version}, nil
}

func (c *HTTPClient) Set(key string, value []byte, ttlSecs ...uint64) (*SetResult, error) {
	return c.setBytes(key, value, nil, optionalUint64FromSlice(ttlSecs...))
}

func (c *HTTPClient) SetString(key, value string, ttlSecs ...uint64) (*SetResult, error) {
	return c.Set(key, []byte(value), ttlSecs...)
}

func (c *HTTPClient) SetInNamespace(key string, value []byte, namespace string, ttlSecs ...uint64) (*SetResult, error) {
	ns, err := resolveCoreNamespaceValue(c.strictMode, "set", key, namespace)
	if err != nil {
		return nil, err
	}
	return c.setBytes(key, value, ns, optionalUint64FromSlice(ttlSecs...))
}

func (c *HTTPClient) SetStringInNamespace(key, value, namespace string, ttlSecs ...uint64) (*SetResult, error) {
	return c.SetInNamespace(key, []byte(value), namespace, ttlSecs...)
}

func (c *HTTPClient) SetNX(key string, value []byte, ttlSecs uint64, namespace ...string) (*SetNXResult, error) {
	ns, err := resolveCoreNamespace(c.strictMode, "set", key, namespace...)
	if err != nil {
		return nil, err
	}
	path := "/key/" + url.PathEscape(key) + "?nx=1"
	if ttlSecs > 0 {
		path += fmt.Sprintf("&ttl=%d", ttlSecs)
	}
	b, status, err := c.request(http.MethodPost, path, value, "application/octet-stream", namespaceHeaderPtr(ns))
	if err != nil {
		return nil, err
	}
	if status == http.StatusBadRequest || status == http.StatusNotFound || status == http.StatusNotImplemented {
		if err := parseAtomicHTTPUnsupported(b, "SET_NX"); err != nil {
			return nil, err
		}
	}
	if err := parseHTTPError(status, b); err != nil {
		return nil, err
	}
	var payload struct {
		Created bool   `json:"created"`
		Version string `json:"version"`
	}
	if err := json.Unmarshal(b, &payload); err != nil {
		return nil, err
	}
	version, err := strconv.ParseUint(payload.Version, 10, 64)
	if err != nil {
		return nil, err
	}
	return &SetNXResult{Created: payload.Created, Version: version}, nil
}

func (c *HTTPClient) Incr(key string, delta int64, ttlSecsOnCreate uint64, namespace ...string) (*IncrResult, error) {
	ns, err := resolveCoreNamespace(c.strictMode, "set", key, namespace...)
	if err != nil {
		return nil, err
	}
	payload := map[string]any{"delta": delta}
	if ttlSecsOnCreate > 0 {
		payload["ttl_secs_on_create"] = ttlSecsOnCreate
	}
	body, _ := json.Marshal(payload)
	b, status, err := c.request(http.MethodPost, "/key/"+url.PathEscape(key)+"/incr", body, "application/json", namespaceHeaderPtr(ns))
	if err != nil {
		return nil, err
	}
	if status == http.StatusBadRequest || status == http.StatusNotFound || status == http.StatusNotImplemented {
		if err := parseAtomicHTTPUnsupported(b, "INCR"); err != nil {
			return nil, err
		}
	}
	if err := parseHTTPError(status, b); err != nil {
		return nil, err
	}
	var response struct {
		Value   string `json:"value"`
		Version string `json:"version"`
	}
	if err := json.Unmarshal(b, &response); err != nil {
		return nil, err
	}
	value, err := strconv.ParseInt(response.Value, 10, 64)
	if err != nil {
		return nil, err
	}
	version, err := strconv.ParseUint(response.Version, 10, 64)
	if err != nil {
		return nil, err
	}
	return &IncrResult{Value: value, Version: version}, nil
}

func (c *HTTPClient) Delete(key string, namespace ...string) (bool, error) {
	ns, err := resolveCoreNamespace(c.strictMode, "delete", key, namespace...)
	if err != nil {
		return false, err
	}
	b, status, err := c.request(http.MethodDelete, "/key/"+url.PathEscape(key), nil, "", namespaceHeaderPtr(ns))
	if err != nil {
		return false, err
	}
	if status == http.StatusNoContent {
		return true, nil
	}
	if status == http.StatusNotFound {
		return false, nil
	}
	if err := parseHTTPError(status, b); err != nil {
		return false, err
	}
	return true, nil
}

func (c *HTTPClient) DeleteByPattern(pattern string, namespace ...string) (*DeleteByPatternResult, error) {
	ns, err := resolvePatternNamespace(c.strictMode, "deleteByPattern", pattern, namespace...)
	if err != nil {
		return nil, err
	}
	payload, _ := json.Marshal(map[string]string{"pattern": pattern})
	b, status, err := c.request(http.MethodPost, "/keys/delete-by-pattern", payload, "application/json", namespaceHeaderPtr(ns))
	if err != nil {
		return nil, err
	}
	if err := parseHTTPError(status, b); err != nil {
		return nil, err
	}
	var out DeleteByPatternResult
	if err := json.Unmarshal(b, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (c *HTTPClient) SetTtlByPattern(pattern string, ttlSecs uint64, namespace ...string) (*SetTtlByPatternResult, error) {
	ns, err := resolvePatternNamespace(c.strictMode, "setTtlByPattern", pattern, namespace...)
	if err != nil {
		return nil, err
	}
	m := map[string]any{"pattern": pattern}
	if ttlSecs > 0 {
		m["ttl_secs"] = ttlSecs
	}
	payload, _ := json.Marshal(m)
	b, status, err := c.request(http.MethodPost, "/keys/ttl-by-pattern", payload, "application/json", namespaceHeaderPtr(ns))
	if err != nil {
		return nil, err
	}
	if err := parseHTTPError(status, b); err != nil {
		return nil, err
	}
	var out SetTtlByPatternResult
	if err := json.Unmarshal(b, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (c *HTTPClient) Stats() (*StatsResult, error) {
	b, status, err := c.request(http.MethodGet, "/stats", nil, "", nil)
	if err != nil {
		return nil, err
	}
	if err := parseHTTPError(status, b); err != nil {
		return nil, err
	}
	var out StatsResult
	if err := json.Unmarshal(b, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func namespaceHeader(namespace ...string) map[string]string {
	if len(namespace) == 0 {
		return nil
	}
	return namespaceHeaderValue(namespace[0])
}

func namespaceHeaderPtr(namespace *string) map[string]string {
	if namespace == nil {
		return nil
	}
	return namespaceHeaderValue(*namespace)
}

func namespaceHeaderValue(namespace string) map[string]string {
	ns := strings.TrimSpace(namespace)
	if ns == "" {
		return nil
	}
	return map[string]string{"X-Ditto-Namespace": ns}
}

func allowInsecureTLSDev() bool {
	value := strings.TrimSpace(os.Getenv(insecureTLSDevEnvGate))
	switch strings.ToLower(value) {
	case "1", "true", "yes", "on":
		return true
	default:
		return false
	}
}

func (c *HTTPClient) setBytes(key string, value []byte, namespace *string, ttl *uint64) (*SetResult, error) {
	if err := validateCoreInputs(c.strictMode, "set", key, namespace); err != nil {
		return nil, err
	}
	path := "/key/" + url.PathEscape(key)
	if ttl != nil {
		path += fmt.Sprintf("?ttl=%d", *ttl)
	}
	b, status, err := c.request(http.MethodPut, path, value, "text/plain", namespaceHeaderPtr(namespace))
	if err != nil {
		return nil, err
	}
	if err := parseHTTPError(status, b); err != nil {
		return nil, err
	}
	var payload struct {
		Version uint64 `json:"version"`
	}
	if err := json.Unmarshal(b, &payload); err != nil {
		return nil, err
	}
	return &SetResult{Version: payload.Version}, nil
}

File: ditto-rust-client/src/http_client.rs
use std::time::Duration;
use std::fs;

use base64::Engine;
use reqwest::{Certificate, Client, Method, StatusCode};
use serde::Deserialize;
use serde_json::json;

use crate::client_internal::{
    atomic_http_unsupported_error_body, is_atomic_unsupported_status, namespace_header_value,
};
use crate::errors::{DittoError, Result};
use crate::types::{
    CounterResult, DeleteByPatternResult, GetResult, SetNxResult, SetResult, SetTtlByPatternResult,
    StatsResult,
};
use crate::validation::{normalized_namespace, validate_core_inputs, validate_pattern_inputs};

#[derive(Debug, Clone)]
pub struct HttpClientOptions {
    pub host: String,
    pub port: u16,
    pub tls: bool,
    pub username: Option<String>,
    pub password: Option<String>,
    pub request_timeout: Duration,
    pub connect_timeout: Duration,
    pub strict_mode: bool,
    pub dev_insecure_tls: bool,
    pub trusted_cert_path: Option<String>,
}

impl Default for HttpClientOptions {
    fn default() -> Self {
        Self {
            host: "localhost".to_string(),
            port: 7778,
            tls: false,
            username: None,
            password: None,
            request_timeout: Duration::from_secs(10),
            connect_timeout: Duration::from_secs(10),
            strict_mode: false,
            dev_insecure_tls: false,
            trusted_cert_path: None,
        }
    }
}

#[derive(Debug, Clone)]
pub struct DittoHttpClient {
    base_url: String,
    client: Client,
    username: Option<String>,
    password: Option<String>,
    strict_mode: bool,
}

impl DittoHttpClient {
    pub fn new(options: HttpClientOptions) -> Result<Self> {
        let scheme = if options.tls { "https" } else { "http" };
        if options.tls && options.dev_insecure_tls {
            return Err(DittoError::Validation(
                "dev_insecure_tls=true is insecure and is no longer supported. Use a trusted certificate configuration instead.".to_string(),
            ));
        }
        let mut builder = Client::builder()
            .connect_timeout(options.connect_timeout)
            .timeout(options.request_timeout)
            .danger_accept_invalid_certs(false);
        if let Some(path) = options
            .trusted_cert_path
            .as_deref()
            .map(str::trim)
            .filter(|value| !value.is_empty())
        {
            let pem = fs::read(path).map_err(|e| {
                DittoError::Validation(format!("failed to read trusted_cert_path {path:?}: {e}"))
            })?;
            let cert = Certificate::from_pem(&pem).map_err(|e| {
                DittoError::Validation(format!("failed to parse trusted_cert_path {path:?}: {e}"))
            })?;
            builder = builder.add_root_certificate(cert);
        }
        let client = builder.build()?;

        Ok(Self {
            base_url: format!("{scheme}://{}:{}", options.host, options.port),
            client,
            username: options.username,
            password: options.password,
            strict_mode: options.strict_mode,
        })
    }

    pub async fn ping(&self) -> Result<bool> {
        let response = self.request(Method::GET, "/ping", None, None).await?;
        if response.status() != StatusCode::OK {
            return Ok(false);
        }
        let body: PingResponse = response.json().await?;
        Ok(body.pong)
    }

    pub async fn get(&self, key: &str, namespace: Option<&str>) -> Result<Option<GetResult>> {
        let namespace = normalized_namespace(self.strict_mode, namespace)?;
        validate_core_inputs(self.strict_mode, "get", key, namespace.as_deref())?;
        let path = format!("/key/{}", url_encode(key));
        let response = self
            .request(Method::GET, &path, namespace.as_deref(), None)
            .await?;
        if response.status() == StatusCode::NOT_FOUND {
            return Ok(None);
        }
        let response = self.assert_ok(response).await?;
        let body: GetResponse = response.json().await?;
        let value = match body.value_base64 {
            Some(value_base64) if !value_base64.is_empty() => base64::engine::general_purpose::STANDARD
                .decode(value_base64.trim())
                .map_err(|e| DittoError::Protocol(format!("invalid value_base64 in HTTP response: {e}")))?,
            _ => body.value.unwrap_or_default().into_bytes(),
        };
        Ok(Some(GetResult {
            value,
            version: body.version,
        }))
    }

    pub async fn set(
        &self,
        key: &str,
        value: impl Into<Vec<u8>>,
        ttl_secs: Option<u64>,
        namespace: Option<&str>,
    ) -> Result<SetResult> {
        let namespace = normalized_namespace(self.strict_mode, namespace)?;
        validate_core_inputs(self.strict_mode, "set", key, namespace.as_deref())?;
        let mut path = format!("/key/{}", url_encode(key));
        if let Some(ttl) = ttl_secs.filter(|ttl| *ttl > 0) {
            path.push_str(&format!("?ttl={ttl}"));
        }
        let body = value.into();
        let response = self
            .request(
                Method::PUT,
                &path,
                namespace.as_deref(),
                Some(("text/plain", body)),
            )
            .await?;
        let response = self.assert_ok(response).await?;
        Ok(response.json().await?)
    }

    pub async fn set_string(
        &self,
        key: &str,
        value: &str,
        ttl_secs: Option<u64>,
        namespace: Option<&str>,
    ) -> Result<SetResult> {
        self.set(key, value.as_bytes().to_vec(), ttl_secs, namespace)
            .await
    }

    /// Atomic create-if-absent. Returns `created=false` (and the existing
    /// version) when the key already exists — no write is performed.
    pub async fn set_nx(
        &self,
        key: &str,
        value: impl Into<Vec<u8>>,
        ttl_secs: Option<u64>,
        namespace: Option<&str>,
    ) -> Result<SetNxResult> {
        let namespace = normalized_namespace(self.strict_mode, namespace)?;
        validate_core_inputs(self.strict_mode, "set", key, namespace.as_deref())?;
        let mut path = format!("/key/{}?nx=1", url_encode(key));
        if let Some(ttl) = ttl_secs.filter(|ttl| *ttl > 0) {
            path.push_str(&format!("&ttl={ttl}"));
        }
        let response = self
            .request(
                Method::POST,
                &path,
                namespace.as_deref(),
                Some(("application/octet-stream", value.into())),
            )
            .await?;
        if is_atomic_unsupported_status(response.status()) {
            let body = response.bytes().await?;
            return Err(atomic_http_unsupported_error_body(&body, "SET_NX"));
        }
        let response = self.assert_ok(response).await?;
        let body: SetNxResponse = response.json().await?;
        Ok(SetNxResult {
            created: body.created,
            version: parse_u64_field(&body.version, "version")?,
        })
    }

    /// Atomic counter increment. Creates the key at `delta` if absent
    /// (with `ttl_secs_on_create`); never resets the TTL of an existing key.
    pub async fn incr(
        &self,
        key: &str,
        delta: i64,
        ttl_secs_on_create: Option<u64>,
        namespace: Option<&str>,
    ) -> Result<CounterResult> {
        let namespace = normalized_namespace(self.strict_mode, namespace)?;
        validate_core_inputs(self.strict_mode, "set", key, namespace.as_deref())?;
        // Send delta as a JSON string so the int64 survives any consumer that
        // would otherwise coerce a large number to a float (server accepts both).
        let payload = if let Some(ttl) = ttl_secs_on_create.filter(|ttl| *ttl > 0) {
            json!({ "delta": delta.to_string(), "ttl_secs_on_create": ttl })
        } else {
            json!({ "delta": delta.to_string() })
        };
        let path = format!("/key/{}/incr", url_encode(key));
        let response = self
            .request(
                Method::POST,
                &path,
                namespace.as_deref(),
                Some(("application/json", serde_json::to_vec(&payload)?)),
            )
            .await?;
        if is_atomic_unsupported_status(response.status()) {
            let body = response.bytes().await?;
            return Err(atomic_http_unsupported_error_body(&body, "INCR"));
        }
        let response = self.assert_ok(response).await?;
        let body: IncrResponse = response.json().await?;
        Ok(CounterResult {
            value: parse_i64_field(&body.value, "value")?,
            version: parse_u64_field(&body.version, "version")?,
        })
    }

    pub async fn delete(&self, key: &str, namespace: Option<&str>) -> Result<bool> {
        let namespace = normalized_namespace(self.strict_mode, namespace)?;
        validate_core_inputs(self.strict_mode, "delete", key, namespace.as_deref())?;
        let path = format!("/key/{}", url_encode(key));
        let response = self
            .request(Method::DELETE, &path, namespace.as_deref(), None)
            .await?;
        match response.status() {
            StatusCode::NO_CONTENT => Ok(true),
            StatusCode::NOT_FOUND => Ok(false),
            _ => {
                let _ = self.assert_ok(response).await?;
                Ok(true)
            }
        }
    }

    pub async fn delete_by_pattern(
        &self,
        pattern: &str,
        namespace: Option<&str>,
    ) -> Result<DeleteByPatternResult> {
        let namespace = normalized_namespace(self.strict_mode, namespace)?;
        validate_pattern_inputs(
            self.strict_mode,
            "deleteByPattern",
            pattern,
            namespace.as_deref(),
        )?;
        let payload = serde_json::to_vec(&json!({ "pattern": pattern }))?;
        let response = self
            .request(
                Method::POST,
                "/keys/delete-by-pattern",
                namespace.as_deref(),
                Some(("application/json", payload)),
            )
            .await?;
        let response = self.assert_ok(response).await?;
        Ok(response.json().await?)
    }

    pub async fn set_ttl_by_pattern(
        &self,
        pattern: &str,
        ttl_secs: Option<u64>,
        namespace: Option<&str>,
    ) -> Result<SetTtlByPatternResult> {
        let namespace = normalized_namespace(self.strict_mode, namespace)?;
        validate_pattern_inputs(
            self.strict_mode,
            "setTtlByPattern",
            pattern,
            namespace.as_deref(),
        )?;
        let payload = if let Some(ttl) = ttl_secs.filter(|ttl| *ttl > 0) {
            serde_json::to_vec(&json!({ "pattern": pattern, "ttl_secs": ttl }))?
        } else {
            serde_json::to_vec(&json!({ "pattern": pattern }))?
        };
        let response = self
            .request(
                Method::POST,
                "/keys/ttl-by-pattern",
                namespace.as_deref(),
                Some(("application/json", payload)),
            )
            .await?;
        let response = self.assert_ok(response).await?;
        Ok(response.json().await?)
    }

    pub async fn stats(&self) -> Result<StatsResult> {
        let response = self.request(Method::GET, "/stats", None, None).await?;
        let response = self.assert_ok(response).await?;
        Ok(response.json().await?)
    }

    async fn request(
        &self,
        method: Method,
        path: &str,
        namespace: Option<&str>,
        body: Option<(&str, Vec<u8>)>,
    ) -> Result<reqwest::Response> {
        let mut request = self
            .client
            .request(method, format!("{}{}", self.base_url, path));
        if let (Some(username), Some(password)) = (&self.username, &self.password) {
            request = request.basic_auth(username, Some(password));
        }
        if let Some(namespace) = namespace_header_value(namespace) {
            request = request.header("X-Ditto-Namespace", namespace);
        }
        if let Some((content_type, body)) = body {
            request = request.header("Content-Type", content_type).body(body);
        }
        Ok(request.send().await?)
    }

    async fn assert_ok(&self, response: reqwest::Response) -> Result<reqwest::Response> {
        if response.status().is_success() {
            return Ok(response);
        }
        let status = response.status();
        let body = response.bytes().await?;
        let mut code = http_status_to_code(status).to_string();
        let mut message = String::from_utf8_lossy(&body).to_string();
        if let Ok(payload) = serde_json::from_slice::<ErrorResponse>(&body) {
            if let Some(error) = payload.error.filter(|value| !value.is_empty()) {
                code = error;
            }
            if let Some(payload_message) = payload.message.filter(|value| !value.is_empty()) {
                message = payload_message;
            } else if message.is_empty() {
                message = code.clone();
            }
        }
        Err(DittoError::server(code, message))
    }
}

fn parse_u64_field(raw: &str, field: &str) -> Result<u64> {
    raw.parse::<u64>()
        .map_err(|_| DittoError::Protocol(format!("invalid {field} in HTTP response: {raw:?}")))
}

fn parse_i64_field(raw: &str, field: &str) -> Result<i64> {
    raw.parse::<i64>()
        .map_err(|_| DittoError::Protocol(format!("invalid {field} in HTTP response: {raw:?}")))
}

fn http_status_to_code(status: StatusCode) -> &'static str {
    match status.as_u16() {
        503 => "NodeInactive",
        504 => "WriteTimeout",
        404 => "KeyNotFound",
        _ => "InternalError",
    }
}

fn url_encode(value: &str) -> String {
    value
        .bytes()
        .flat_map(|byte| match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                vec![byte as char]
            }
            _ => format!("%{byte:02X}").chars().collect(),
        })
        .collect()
}

#[derive(Deserialize)]
struct PingResponse {
    pong: bool,
}

#[derive(Deserialize)]
struct GetResponse {
    value: Option<String>,
    value_base64: Option<String>,
    version: u64,
}

#[derive(Deserialize)]
struct ErrorResponse {
    error: Option<String>,
    message: Option<String>,
}

#[derive(Deserialize)]
struct SetNxResponse {
    created: bool,
    version: String,
}

#[derive(Deserialize)]
struct IncrResponse {
    value: String,
    version: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dev_insecure_tls_is_rejected() {
        let err = DittoHttpClient::new(HttpClientOptions {
            tls: true,
            dev_insecure_tls: true,
            ..HttpClientOptions::default()
        })
        .unwrap_err();
        assert!(err.to_string().contains("no longer supported"));
    }

    #[test]
    fn trusted_cert_path_defaults_to_none() {
        let options = HttpClientOptions::default();
        assert!(options.trusted_cert_path.is_none());
    }
}
